ksqldb-engine/src/main/java/io/confluent/ksql/internal/UtilizationMetricsListener.java

```diff
@@ -0,0 +1,192 @@
+                    .filter(m -> m.metricName().group().equals(STREAM_METRICS) &&
+                            m.metricName().tags().getOrDefault(THREAD_ID, "").equals(threadName) &&
+                            metrics.contains(m.metricName().name()))
+                    .collect(Collectors.toMap(k -> k.metricName().name(), v -> (Double) v.metricValue()));
+
+            final List<Metric> startTimeList = allMetrics.stream()
+                    .filter(m -> m.metricName().group().equals(STREAM_METRICS) &&
+                            m.metricName().tags().getOrDefault(THREAD_ID, "").equals(threadName) &&
+                            m.metricName().name().equals("thread-start-time"))
+                    .collect(Collectors.toList());
+            final Long threadStartTime = startTimeList.size() != 0 ? (Long) startTimeList.get(0).metricValue() : 0L;
+
+            double blockedTime = 0.0;
+            if (threadStartTime > windowStart) {
+                blockedTime += threadStartTime - windowStart;
+            }
+            final double newFlushTime = threadMetrics.getOrDefault("flush-time-total", windowSize);
+            double newConsumerIOTime = consumerMetrics.getOrDefault("io-waittime-total", windowSize)
+                + consumerMetrics.getOrDefault("iotime-total", windowSize)
+                + restoreConsumerMetrics.getOrDefault("io-waittime-total", windowSize)
+                + restoreConsumerMetrics.getOrDefault("iotime-total", windowSize);
+            newConsumerIOTime = newConsumerIOTime / (1000 * 1000);
+
+            double newProducerBufferBlockTime = producerMetrics.getOrDefault("bufferpool-wait-time-total", windowSize);
+            newProducerBufferBlockTime = newProducerBufferBlockTime / (1000 * 1000);
+
+            blockedTime += Math.max(newFlushTime - previousFlushTime.getOrDefault(threadName, 0.0), 0);
+            previousFlushTime.put(threadName, newFlushTime);
+
+            blockedTime += Math.max(newConsumerIOTime - previousConsumerIOTime.getOrDefault(threadName, 0.0), 0);
+            previousConsumerIOTime.put(threadName, newConsumerIOTime);
+
+            blockedTime += Math.max(newProducerBufferBlockTime - previousProducerBufferWaitTime.getOrDefault(threadName, 0.0), 0);
+            previousProducerBufferWaitTime.put(threadName, newProducerBufferBlockTime);
+
+            blockedTime = Math.min(windowSize, blockedTime);
+            return blockedTime;
+
+        } catch (ClassCastException e) {
+            logger.error("Class cast exception in `UtilizationMetricsListener`. The underlying" +
+                    "streams metrics might have changed type." + e.getMessage());
+            return 0.0;
+        } catch (Exception e) {
+            logger.error("oops: ", e);
+            throw e;
+        }
+    }
+
+    private void initializeMetrics() {
+        metrics.add("io-waittime-total");
+        metrics.add("iotime-total");
+        metrics.add("bufferpool-wait-time-total");
```

Comment:
Besides this one in producer, all other metrics seem unfamiliar to me. Could you point to me where do you find them, plus `thread-start-time`?